ports/esp32/README.md

```diff
@@ -26,59 +26,122 @@ There are two main components that are needed to build the firmware:
-for the ESP-IDF instead of the Windows ones.
-
-You will also need either Python 2 or Python 3, along with the `pyserial` and
-`pyparsing` packages installed for the version of Python that you will be using
-(when building you can use, eg, `make PYTHON=python2` to specify the version
-used).  To install the required packages do:
+If you are on a Windows machine then the [Windows Subsystem for
+Linux](https://msdn.microsoft.com/en-au/commandline/wsl/install_guide) is the
+most efficient way to install the ESP32 toolchain and build the project. If
+you use WSL then follow the Linux instructions rather than the Windows
+instructions.
+
+The ESP-IDF changes quickly and MicroPython only supports certain versions.
+The git hash of these versions (one for 3.x, one for 4.x) can be found by
+running `make` without a configured `ESPIDF`. Then you can fetch the
+required IDF using the following command:
+
+```bash
+$ cd ports/esp32
+$ make ESPIDF=
+$ cd $HOME/src/github.com/espressif  # Or any path you like
+$ git clone https://github.com/espressif/esp-idf.git
+$ git checkout <Current supported ESP-IDF commit hash>
+$ git submodule update --init --recursive
+```
+
+Note: The ESP IDF v4.x support is currently experimental. It does not
+currently support PPP or wired Ethernet.
+
+Python dependencies
+===================
+
+You will also need other dependencies from the IDF, see
+`esp-idf/requirements.txt`, but at a minimum you need `pyserial>=3.0` and
+`pyparsing>=2.0.3,<2.4.0`.
+
+You can use Python 2 or Python 3, use `make PYTHON=python2` if you need to
+override the system default.
+
+It is recommended to use a Python virtual environment. Even if your system
+package manager already provides these libraries, the IDF v4.x is currently
+incompatible with pyparsing 2.4 and higher.
+
+For example, to set up a Python virtual environment from scratch:
+
+```bash
+cd ports/esp32
+python3 -m venv build-venv
+source build-venv/bin/activate
+pip install --upgrade pip
+pip install path/to/esp-idf/requirements.txt
+```
+
+Subsequent times will only require the first two lines.
+
+Then, to install the toolchain (which includes the GCC compiler, linker, binutils,
+etc), there are two options:
+
+1. Using the IDF scripts to install the toolchain (IDF 4.x only)
+================================================================
+
+Follow the steps at the [Espressif Getting Started guide](https://docs.espressif.com/projects/esp-idf/en/v4.0/get-started/index.html#step-3-set-up-the-tools).
+
+This will guide you through using the `install.sh` (or `install.bat`) script
+to download the toolchain and add it to your `PATH`. The steps are summarised
+below:
+
+After you've cloned and checked out the IDF to the correct version (see
+above), run the `install.sh` script:
+
+```bash
+$ cd path/to/esp-idf
+$ ./install.sh    # (or install.bat on Windows)
+```
+
+Then the `ports/esp32` directory, source the `export.sh` script to set up the
```

Comment:
"The the"?
Also, I'm not following, why does one need to run export.sh from within the ports/esp32 directory?

Reply:
You source it, not run it. It sets PATH.
But you're right of course that it can be run from anywhere. Have clarified this slightly.